mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -0,0 +1,530 @@
+                            "entity": token_text,
+                            "raw_entity": raw_token["text"],
+                            "raw_token_index": i,
+                            "raw_start": raw_token["start"],
+                        }
+                    )
+        return normalized_tokens
+
+    def _normalize_single_token(self, text):
+        """Normalize an individual token by processing text with all normalizers.
+
+        Args:
+            text (str): Text to normalize.
+        Returns:
+            normalized_texts (List[str]): Normalized texts. We tokenize the post-norm text and
+                split the entity if possible.
+        """
+        normalized_text = text
+        for normalizer in self.normalizers:
+            normalized_text = TextPreparationPipeline.modify_around_annotations(
+                text=normalized_text, function=normalizer.normalize
+            )
+        # We tokenize the post-norm text and split the entity if possible
+        # Ex: normalize("o'clock") -> "o clock" -> ["o", "clock"]
+        return [t["text"] for t in self.tokenize(normalized_text)]
+
+    def tokenize(self, text):
+        """
+        Args:
+            text (str): Input text.
+        Returns:
+            tokens (List[str]): List of tokens.
+        """
+        return TextPreparationPipeline.tokenize_around_mindmeld_annotations(
+            text=text, function=self.tokenizer.tokenize
+        )
+
+    def stem_word(self, word):
+        """
+        Gets the stem of a word. For example, the stem of the word 'fishing' is 'fish'.
+
+        Args:
+            words (List[str]): List of words to stem.
+
+        Returns:
+            stemmed_words (List[str]): List of stemmed words.
+        """
+        return self.stemmer.stem_word(word)
+
+    @staticmethod
+    def find_mindmeld_annotation_re_matches(text):
+        """
+        Args:
+            text (str): The string to find mindmeld annotation instances
+                (" {entity_text|entity_type} ")
+        Returns:
+            matches (List[sre.SRE_Match object]): Regex match objects.
+        """
+        return list(MINDMELD_ANNOTATION_PATTERN.finditer(text))
+
+    @staticmethod
+    def modify_around_annotations(text, function):
+        """Applied a function around the mindmeld annotation.
+
+        function(pre_entity_text) + { + function(entity_text) + |entity_name}
+            + function(post_entity_text)
+
+        Args:
+            text (str): Original sentence with markup to modify.
+            function (function): Function to apply around the annotation
+        Returns:
+            modified_text (str): Text modified around annotations.
+        """
+        matches = TextPreparationPipeline.find_mindmeld_annotation_re_matches(text)
+
+        modified_text = []
+        prev_entity_end = 0
+
+        for match in matches:
+            entity_start, entity_end = match.span()
+
+            # Adds "function(pre_entity_text)" {..
```

Comment:
Shouldnt this be `# Adds "function(pre_entity_text)"{..`

Reply:
Good catch, fixed